textPostConfig and photoPostConfig templates [pr]

aaronschachter opened the pull request:
#### What's this PR do?

Returns the field values from a botConfig's `postConfig` entry as properties in the `templates` object of a `GET /campaigns/:id` response.

Contentful changes:

*  [x] Added a new `photoPostConfig` content type, with `photoPost` fields like `gambitSignupMenuMessage`, `askQuantityMessage`

* [x] Added `askTextMessage`, `invalidTextMessage`, and `menuCompletedMessages` fields to the `textPostConfig` content type.

* [x] Added `photoPostConfig` as an available type to reference as a botConfig's `postConfig` field.

* [x] Create `photoPostConfig` entries for all botConfigs currently running on staging and production, migrating overrides. This sounds like the stuff Contentful migrations are made for, but the number's small enough that I'll just manually add 💻 📝 Campaigns id's to edit:

    * [x] 7 - Mirror Messages
    * [x] 2178  - Give A Spit (thirdParty)
    * [x] 7059 - VCard (thirdParty)
    * [x] 7483 - Rinse Recycle Repeat
    * [x] 7951 - Gun Violence (thirdParty)
    * [x] 8017 - Grab the Mic (thirdParty)
    * [x] 8038 - Save Mascots
<br>

* [x] Make `postConfig` field mandatory 

After go live:

* [ ] Remove botConfig fields like `askQuantity` that have been migrated to the photoPostConfig type.

Gambit changes:

* Renames the botConfig `config.templates` as `config.templatesByContentType`, adds `textPostConfig` and `photoPostConfig` properties (#1032) per Contentful changes above 👆 

* Refactors the `botConfig` helper to gather templates based on content type

* Adds new `askText` and `invalidText` replies for use in Text Post middleware

#### How should this be reviewed?

Upon staging deploy:

* Verify textPost conversation flow (LUCKYBOT)

* Verify photoPost conversation flow (BOOKBOT)

#### Relevant tickets
#1021 

#### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Tested on staging.
- [x] Deploy https://github.com/DoSomething/gambit-conversations/pull/317 to Conversations prod before deploying this branch to Campaigns prod

aaronschachter commented on lib/middleware/campaigns/single/bot-config-parse.js:
```diff
@@ -15,9 +15,6 @@ module.exports = function parseBotConfig() {
       req.campaign.botConfig.templates = templates;
-      // Adding this for backwards-compatability.
-      // TODO: remove this property once we make change on Conversations side to inspect botConfig.
-      req.campaign.templates = templates;
```
Done in https://github.com/DoSomething/gambit-conversations/pull/313

aaronschachter commented on config/lib/helpers/botConfig.js:
```diff
@@ -23,11 +21,45 @@ module.exports = {
+      },
+      invalidAskSignupResponse: {
+        fieldName: 'invalidSignupResponseMessage',
+        default: `${defaultText.invalidInput} Did you want to join {{title}}${defaultText.yesNo}`,
```
It should! Good 👀

aaronschachter commented on lib/helpers/botConfig.js:
```diff
@@ -15,56 +15,85 @@ function fetchByCampaignId(campaignId) {
-function getDefaultTextForBotConfigTemplateName(templateName) {
-  return config.templates.botConfig[templateName].default;
+function getDefaultValueFromEntryAndTemplateName(entry, templateName) {
+  const templateConfig = module.exports
```
I'll revisit this to double check -- I seemed to be having trouble in some of the unit tests without it.

aaronschachter commented:
Backfilled all current production campaigns with `photoPostConfig` entries, and added staging keywords for them when they didn't exist. I'll test each campaign flow and check off the "tested on staging" item when complete.

@rapala61 would you mind giving a quick 👀 when you have a chance?

aaronschachter commented:
After discussing in Slack, the `textPost`configuration needs tweaks -- I forgot to account for the signup confirmation (the `externalSignupMenu`) we send for web signups via SMS, where we want to send a web signup confirmation asking a user to text back a reportback for Text Reportbacks.

I think the quick win to account for it is to use `signupMenu` and `externalSignupMenu` and `invalidSignupMenuResponse` command fields instead of the `askText`, `invalidText`. The Signup Menu for a Text Reportback would just be simply to respond with the text post. Will update this branch and close https://github.com/DoSomething/gambit-conversations/pull/314 accordingly.

aaronschachter commented:
Another idea is to keep as is but add something like an `externalAskText`, or `webSignupAskText` field to use in place of a `externalSignupMenu` template... and check the `postType` of the campaign in our Signup Message route. If it's set to `text`, send an `externalAskText` if not, send `externalSignupMenu`.. .

aaronschachter commented:
Backtracking per Slack thread - web signup confirmation messages are nice to have but not required. It may be cleanest to just not send Signup messages if a botConfig is configured with a `textPostConfig`.

rapala61 commented:
I'm getting an error in G-Conversations:

I think it might be due to the missing `invalidCompletedMenuCommand` template in the response from G-Campaigns.

### Log Screenshot 

<img width="1429" alt="screen shot 2018-04-10 at 11 05 14 am" src="https://user-images.githubusercontent.com/863301/38568807-3c83662c-3caf-11e8-845b-b1abb01656d0.png">

aaronschachter commented on lib/helpers/botConfig.js:
```diff
@@ -15,56 +15,85 @@ function fetchByCampaignId(campaignId) {
+ * @return {String}
+ */
+function getFieldValueFromEntryAndTemplateName(entry, templateName) {
+  const templateConfig = module.exports
```
Agreed, maybe rename as `templateData` ?

aaronschachter commented on lib/helpers/botConfig.js:
```diff
@@ -15,56 +15,85 @@ function fetchByCampaignId(campaignId) {
+ * @return {Object}
  */
-function getTemplatesFromBotConfig(botConfig) {
+function getTemplatesFromEntry(entry) {
```
I was debating on using `FromContentfulEntry` instead of `Entry,`, would that help?  My hesitation for renaming to `FromConfig` is that it seems unclear whether its a contentful entry like a `botConfig`, `postTextConfig`, or `photoPostConfig` or a file in the root `config` dir.

aaronschachter commented:
I unpublished the keyword for 2070 last night from staging because it was one less `photoPostConfig` entry to set up -- your user must have been in a conversation for it. This does bring up a minor bug, where a user may be in a campaign that no longer has keywords published -- they don't get the closed message. In practice, ideally what should happen is the keyword is published long enough while campaign has been closed, so user gets the closed message.  Definitely room for improvement though.

I don't mind republishing/updating 2070 back to staging as a quick fix.

aaronschachter commented:
Thinking through templates names of a `textPostConfig`. 


Currently they are named `askText`, `invalidText` and `completedMenu` --  where the `completedMenu` is asking user to text another reportback (the `completedMenu` of a `photoPostConfig` prompts user to start another post, otherwise says I didn't get that). Outstanding tasks:

* In Contentful, add a new value for `askText` on the broadcast template field to close https://github.com/DoSomething/gambit-conversations/issues/304. A potential gotcha here is an editor could set the template to `askText` for a campaign that doesn't have a `textPostConfig`, which makes me consider a `broadcastConfig` type of field for conditional fields like we're doing for `botConfig`'s... 

* In the Conversations Signup message middleware, return a 204 if a campaign.botConfig.postType is set to `'text'`, like how we return a 204 for no keywords. Alternatively, if we wanted to send a web signup confirmation message, we'd need a new template -- something like `webSignupAskText`, `externalSignupAskText` which feels a little gross

This brings to mind using `gambitSignupMenu` and `externalSignupMenu` as template names for a `textPhotoConfig`, because we'd get some stuff for free:

* Don't need a new `askText` value for broadcast template, this works out of the box

* Don't need to alter the Signup message middleware either if we want to send the confirmation. Per @classicfreddie's feedback on the feature: It's a nice to have. I'm a little weary of the UX, user signs up via web and sees a textbox, we then send a text asking them to submit a tip instead of using web input

But poses some weirdness. There's still a use for the `invalidText` template, it makes more sense than calling it a `invalidSignupMenuCommand`... as we're not asking for a command like START but text to use for reportback.

aaronschachter commented:
In standup today, we chatted about the things I've been thinking through in this PR around the  `textPostConfig` field/template names... which ultimately surfaced a need for a spec on Gambit message templates. I've taken a [first stab at one here](https://docs.google.com/document/d/1S6ACY0_0AMQ8dnqRsfvfo5vxqdviZXMmhQP1T7WLea4/edit). I like Google docs for specifications for features like starting comment threads around specific highlighted text, or assigning comments to users, etc.

To our pullrequest friends - not sure if it's within your wheelhouse to review technical specs as well as pull requests, but for sure we'd love to get your expertise 👀, I've made the doc public for viewing and commenting if interested.

aaronschachter commented:
I've updated this branch per the spec and our discussion today in story time, deployed to staging in tandem with updates to https://github.com/DoSomething/gambit-conversations/pull/314#issuecomment-380607230. The textPostConfig fields have been renamed (and also added a `webSignupConfirmed` to support confirmation messaging for web signups), but this branch does not renaming any of the photoPostConfig fields for the sake of slowly rolling out these Contentful changes.

aaronschachter commented:
This has been deployed to production. I've deleted the deprecated fields from our `campaign` content type which are now sourced from the `photoPostConfig` type, e.g `askQuantityMessage`, `askCaptionMessage`, etc.
